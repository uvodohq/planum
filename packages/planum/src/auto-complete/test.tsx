import {
  autoUpdate,
  size,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useListNavigation,
  useRole,
} from '@floating-ui/react-dom-interactions'
import React, { forwardRef, useLayoutEffect, useRef, useState } from 'react'

interface ItemProps {
  children: React.ReactNode
  active: boolean
}

const Item = forwardRef<
  HTMLDivElement,
  ItemProps & React.HTMLProps<HTMLDivElement>
>(({ children, active, ...rest }, ref) => {
  const id = useId()
  return (
    <div
      ref={ref}
      role="option"
      id={id}
      aria-selected={active}
      {...rest}
      style={{
        background: active ? 'lightblue' : 'none',
        padding: 4,
        cursor: 'default',
        ...rest.style,
      }}>
      {children}
    </div>
  )
})
Item.displayName = 'Item'

export function AutoCompleteTest({ data }) {
  const [open, setOpen] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const listRef = useRef<Array<HTMLElement | null>>([])

  const { x, y, reference, floating, strategy, context, refs } =
    useFloating<HTMLInputElement>({
      whileElementsMounted: autoUpdate,
      open,
      onOpenChange: setOpen,
      middleware: [
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`,
              maxHeight: `${200}px`,
            })
          },
          padding: 10,
        }),
      ],
    })

  useLayoutEffect(() => {
    // IMPORTANT: When the floating element first opens, this runs when the
    // styles have **not yet** been applied to the element. This can cause an
    // infinite loop as `size` has not yet limited the maxHeight, so the whole
    // page tries to scroll. We must wrap it in rAF.
    requestAnimationFrame(() => {
      if (activeIndex != null) {
        listRef.current[activeIndex]?.scrollIntoView({ block: 'nearest' })
      }
    })
  }, [activeIndex])

  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [
      useRole(context, { role: 'listbox' }),
      useDismiss(context),
      useListNavigation(context, {
        listRef,
        activeIndex,
        onNavigate: setActiveIndex,
        virtual: true,
        loop: true,
        focusItemOnHover: false,
      }),
    ],
  )

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.target.value
    setInputValue(value)

    if (value) {
      setOpen(true)
      setActiveIndex(0)
    } else {
      setOpen(false)
    }
  }

  const items = data.filter((item) =>
    item.toLowerCase().startsWith(inputValue.toLowerCase()),
  )

  return (
    <>
      <input
        {...getReferenceProps({
          ref: reference,
          onChange,
          value: inputValue,
          placeholder: 'Enter fruit',
          'aria-autocomplete': 'list',
          onKeyDown(event) {
            if (
              event.key === 'Enter' &&
              activeIndex != null &&
              items[activeIndex]
            ) {
              setInputValue(items[activeIndex])
              setActiveIndex(null)
              setOpen(false)
            }

            if (event.key === 'Tab') {
              setOpen(false)
            }
          },
        })}
      />
      {open && (
        <div
          {...getFloatingProps({
            ref: floating,
            style: {
              position: strategy,
              left: x ?? 0,
              top: y ?? 0,
              background: '#eee',
              color: 'black',
              overflowY: 'auto',
            },
          })}>
          {items.map((item, index) => (
            // eslint-disable-next-line react/jsx-key
            <Item
              {...getItemProps({
                key: item,
                ref(node) {
                  listRef.current[index] = node
                },
                onClick() {
                  setInputValue(item)
                  setOpen(false)
                  refs.reference.current?.focus()
                },
              })}
              active={activeIndex === index}>
              {item}
            </Item>
          ))}
        </div>
      )}
    </>
  )
}
