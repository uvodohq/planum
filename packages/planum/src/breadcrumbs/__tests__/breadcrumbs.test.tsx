import { render, screen } from '../../test'
import { BreadcrumbItem, Breadcrumbs } from '..'

describe('Breadcrumbs', async () => {
  it('should render 3 item', () => {
    render(
      <Breadcrumbs>
        <BreadcrumbItem>Item 1</BreadcrumbItem>
        <BreadcrumbItem>Item 2</BreadcrumbItem>
        <BreadcrumbItem>Item 3</BreadcrumbItem>
      </Breadcrumbs>,
    )

    expect(screen.getAllByRole('listitem')).toHaveLength(3)
    expect(screen.getByRole('navigation')).toMatchSnapshot()
  })
})
