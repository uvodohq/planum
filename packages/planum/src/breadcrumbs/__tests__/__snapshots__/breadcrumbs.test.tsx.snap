// Vitest Snapshot v1

exports[`Breadcrumbs > should render 3 item 1`] = `
<nav
  aria-label="Breadcrumbs"
>
  <ol
    class="c-dFneIQ"
  >
    <li
      class="c-gUpUYw"
    >
      <span
        class="c-iPYhUp c-kLRYwF c-ijlUVt c-ijlUVt-lazxZu-isCurrent-false"
      >
        Item 1
      </span>
      <span
        class="c-fAlSmZ"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.47 2.47a.75.75 0 0 1 1.06 0l5 5a.75.75 0 0 1 0 1.06l-5 5a.75.75 0 0 1-1.06-1.06L9.94 8 5.47 3.53a.75.75 0 0 1 0-1.06Z"
            fill="#C7C7C7"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </li>
    <li
      class="c-gUpUYw"
    >
      <span
        class="c-iPYhUp c-kLRYwF c-ijlUVt c-ijlUVt-lazxZu-isCurrent-false"
      >
        Item 2
      </span>
      <span
        class="c-fAlSmZ"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.47 2.47a.75.75 0 0 1 1.06 0l5 5a.75.75 0 0 1 0 1.06l-5 5a.75.75 0 0 1-1.06-1.06L9.94 8 5.47 3.53a.75.75 0 0 1 0-1.06Z"
            fill="#C7C7C7"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </li>
    <li
      class="c-gUpUYw"
    >
      <span
        aria-current="page"
        aria-disabled="true"
        class="c-iPYhUp c-kLRYwF c-ijlUVt c-ijlUVt-cLNiCz-isCurrent-true"
      >
        Item 3
      </span>
    </li>
  </ol>
</nav>
`;
